doc/config.rst
``directory_list``
------------------
List of dictionary directory paths (absolute or relative to home)

Default value: ``[".stardict/dic"]``

``style``
---------
Path to application stylesheet file (.qss)

Default value: ``""``

``article_style``
-----------------
Path to article stylesheet file (.css)

Default value: ``""``

``soundex_words_file``
----------------------
Soundex words file

Default value: ``""``

``font_family``
---------------
Application font family

Default value: ``""``

``font_size``
-------------
Application font size

Default value: ``0``

``search_on_type``
------------------
Enable/disable search-on-type

Default value: ``false``

``search_on_type_min_length``
-----------------------------
Minimum query length for search-on-type

Default value: ``3``

``search_on_type_on_regex``
---------------------------
Enable/disable search-on-type in Regex mode

Default value: ``false``

``header_template``
-------------------
HTML template for header (dict name + entry terms)

Default value: ``"<b><font color='#55f'>{{.DictName}}</font></b>\n<font color='#777'> [Score: %{{.Score}}]</font>\n{{if .ShowTerms }}\n<span dir=\"ltr\" style=\"font-size: xx-large;font-weight:bold;\">\n{{ index .Terms 0 }}\n</span>\n{{range slice .Terms 1}}\n<span dir=\"ltr\" style=\"font-size: large;font-weight:bold;\">\n\t<span style=\"color:#ff0000;font-weight:bold;\"> │ </span>\n\t{{ . }}\n</span>\n{{end}}\n{{end}}"``

``header_word_wrap``
--------------------
Enable word-wrapping for header (dict name + entry terms)

Default value: ``true``

``history_disable``
-------------------
Disable history

Default value: ``false``

``history_auto_save``
---------------------
Auto-save history on every new record

Default value: ``true``

``history_max_size``
--------------------
Maximum size for history

Default value: ``100``

``most_frequent_disable``
-------------------------
Disable keeping Most Frequent queries

Default value: ``false``

``most_frequent_auto_save``
---------------------------
Auto-save Most Frequent queries

Default value: ``true``

``most_frequent_max_size``
--------------------------
Maximum size for Most Frequent queries

Default value: ``100``

``favorite_button_image``
-------------------------
Image file name for favorite button. Try favorite-blue-64.png for color blindness

Default value: ``"favorite-green-64.png"``

``favorite_button_hue``
-----------------------
Color hue for favorite button in scan popup (120=green, 240=blue, 0=red)

Default value: ``120``

``max_results_total``
---------------------
Maximum number of search results

Default value: ``40``

``result_tree``
---------------
Use a tree (instead of list) to show results (terms as children)

Default value: ``false``

``random_favorite_search_mode``
-------------------------------
Search mode for Random Favorite

Default value: ``"wordMatch"``

``start_hidden``
----------------
Hide main window on startup (if tray icon is available)

Default value: ``false``

``desktop_widget``
------------------
Desktop Widget: enable

Default value: ``true``

``desktop_widget_click_time``
-----------------------------
Desktop Widget: max click time in millisecond

Default value: ``100``

``scan_popup_clipboard``
------------------------
Scan Popup: activate on copy to clipboard

Default value: ``false``

``scan_popup_selection``
------------------------
Scan Popup: activate on selection

Default value: ``false``

``scan_popup_api``
------------------
Scan Popup: activate via API (socket/pipe)

Default value: ``true``

``scan_popup_mode``
-------------------
Scan Popup: search mode

Default value: ``"fuzzy"``

``scan_popup_min_score``
------------------------
Scan Popup: minimum score (0 to 100)

Default value: ``0``

``scan_popup_width``
--------------------
Scan Popup: window width

Default value: ``700``

``scan_popup_height``
---------------------
Scan Popup: window height

Default value: ``400``

``scan_popup_max_count``
------------------------
Scan Popup: max number of windows

Default value: ``3``

``scan_popup_history``
----------------------
Scan Popup: add to history

Default value: ``true``

``scan_popup_header_icons``
---------------------------
Scan Popup: use icons for header buttons

Default value: ``true``

``scan_popup_header_template``
------------------------------
Scan Popup: HTML template for header (dict name and score)

Default value: ``"<span style='font-weight:200;'>{{.DictName}} [Score: %{{.Score}}]</span>"``

``scan_popup_terms_style``
--------------------------
Scan Popup: CSS style string for terms

Default value: ``"style=\"font-size:large;font-weight:bold;\""``

``scan_popup_font_size_factor``
-------------------------------
Scan Popup: font size factor (relative to app)

Default value: ``0.8``

``scan_popup_bypass_window_manager``
------------------------------------
Scan Popup: bypass window manager

Default value: ``true``

``random_favorite_popup_interval_seconds``
------------------------------------------
Show a random favorite term popup every N seconds (0 to disable)

Default value: ``0``

``audio``
---------
Enable audio in article

Default value: ``true``

``audio_mpv``
-------------
Use ``mpv`` command for playing audio

Default value: ``false``

``audio_download_timeout``
--------------------------
Timeout for downloading audio files

Default value: ``"1s"``

``audio_auto_play``
-------------------
Number of audio file to auto-play, set ``0`` to disable.

Default value: ``1``

``audio_auto_play_wait_between``
--------------------------------
Wait time between multiple audio files on auto-play

Default value: ``"800ms"``

``audio_auto_play_min_score``
-----------------------------
Minimum score for audio auto-play

Default value: ``100``

``audio_volume``
----------------
Volume for playing audio, 0 to 100 (% multiplied by dict-specific volume)

Default value: ``70``

``embed_external_stylesheet``
-----------------------------
Embed external stylesheet/css in article

Default value: ``false``

``resource_http_download_timeout``
----------------------------------
Timeout for downloading http/https resources in article

Default value: ``"2s"``

``color_mapping``
-----------------
Mapping for colors used in article

Default value: ``{}``

``loading_popup_style_str``
---------------------------
Stylesheet (text) for 'Loading' popup

Default value: ``"border: 1px solid red; background-color: #333; color: white"``

``article_zoom_factor``
-----------------------
Zoom factor for article with mouse wheel or keyboard

Default value: ``1.1``

``article_arrow_keys``
----------------------
Use arrow keys to scroll through article (when focused)

Default value: ``false``

``misc_buttons_vertical_padding``
---------------------------------
Misc buttons vertical padding

Default value: ``5``

``reduce_minimum_window_width``
-------------------------------
Use smaller buttons to reduce minimum width of window

Default value: ``false``

``local_server_ports``
----------------------
Ports for local server. Server runs on first port; Client tries all

Default value: ``["8357"]``

``local_client_timeout``
------------------------
Timeout for local web client

Default value: ``"100ms"``

``web_enable``
--------------
Set true/false and restart to enable/disable web service & web app

Default value: ``false``

``web_expose``
--------------
Expose web service & web app to outside (otherwise only available to 127.0.0.1)

Default value: ``false``

``web_search_on_type``
----------------------
Web: Enable/disable search-on-type

Default value: ``false``

``web_search_on_type_min_length``
---------------------------------
Web: Minimum query length for search-on-type

Default value: ``3``

``web_search_on_type_on_regex``
-------------------------------
Web: Enable/disable search-on-type in Regex mode

Default value: ``false``

``web_show_powered_by``
-----------------------
Show 'Powered By ...' footer in web.

Default value: ``true``

``search_worker_count``
-----------------------
The number of workers / goroutines used for search

Default value: ``8``

``search_timeout``
------------------
Timeout for search on each dictionary. Only works if ``search_worker_count > 1``

Default value: ``"5s"``

``logging.no_color``
--------------------
Disable log colors

Default value: ``false``

``logging.level``
-----------------
Log level

Default value: ``"info"``



pkg/config/config.go
package config

import (
	"errors"
	"log/slog"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/BurntSushi/toml"
)

const fileName = "config.toml"

var mutex sync.Mutex

var PrivateMode = false

func init() {
	dir := GetConfigDir()
	err := os.MkdirAll(dir, 0o755)
	if err != nil {
		slog.Error("error in MkdirAll: " + err.Error())
	}
}

type LoggingConfig struct {
	NoColor bool   `toml:"no_color" doc:"Disable log colors"`
	Level   string `toml:"level" doc:"Log level"`
}

type Config struct {
	Logging LoggingConfig `toml:"logging" doc:"Logging config"`

	DirectoryList []string `toml:"directory_list" doc:"List of dictionary directory paths (absolute or relative to home)"`

	Style string `toml:"style" doc:"Path to application stylesheet file (.qss)"`

	ArticleStyle string `toml:"article_style" doc:"Path to article stylesheet file (.css)"`

	SoundexWordsFile string `toml:"soundex_words_file" doc:"Soundex words file"`

	FontFamily string `toml:"font_family" doc:"Application font family"`
	FontSize   int    `toml:"font_size" doc:"Application font size"`

	SearchOnType          bool `toml:"search_on_type" doc:"Enable/disable search-on-type"`
	SearchOnTypeMinLength int  `toml:"search_on_type_min_length" doc:"Minimum query length for search-on-type"`
	SearchOnTypeOnRegex   bool `toml:"search_on_type_on_regex" doc:"Enable/disable search-on-type in Regex mode"`

	HeaderTemplate string `toml:"header_template" doc:"HTML template for header (dict name + entry terms)"`
	HeaderWordWrap bool   `toml:"header_word_wrap" doc:"Enable word-wrapping for header (dict name + entry terms)"`

	HistoryDisable  bool `toml:"history_disable" doc:"Disable history"`
	HistoryAutoSave bool `toml:"history_auto_save" doc:"Auto-save history on every new record"`
	HistoryMaxSize  int  `toml:"history_max_size" doc:"Maximum size for history"`

	MostFrequentDisable  bool `toml:"most_frequent_disable" doc:"Disable keeping Most Frequent queries"`
	MostFrequentAutoSave bool `toml:"most_frequent_auto_save" doc:"Auto-save Most Frequent queries"`
	MostFrequentMaxSize  int  `toml:"most_frequent_max_size" doc:"Maximum size for Most Frequent queries"`

	FavoriteButtonImage string `toml:"favorite_button_image" doc:"Image file name for favorite button. Try favorite-blue-64.png for color blindness"`

	FavoriteButtonHue int `toml:"favorite_button_hue" doc:"Color hue for favorite button in scan popup (120=green, 240=blue, 0=red)"`

	MaxResultsTotal int `toml:"max_results_total" doc:"Maximum number of search results"`

	ResultTree bool `toml:"result_tree" doc:"Use a tree (instead of list) to show results (terms as children)"`

	RandomFavoriteSearchMode string `toml:"random_favorite_search_mode" doc:"Search mode for Random Favorite"`

	StartHidden bool `toml:"start_hidden" doc:"Hide main window on startup (if tray icon is available)"`

	DesktopWidget bool `toml:"desktop_widget" doc:"Desktop Widget: enable"`
	// DesktopWidgetBypassWindowManager bool `toml:"desktop_widget_bypass_window_manager" doc:"Desktop Widget: bypass window manager"`

	DesktopWidgetClickTime int `toml:"desktop_widget_click_time" doc:"Desktop Widget: max click time in millisecond"`

	ScanPopupClipboard bool   `toml:"scan_popup_clipboard" doc:"Scan Popup: activate on copy to clipboard"`
	ScanPopupSelection bool   `toml:"scan_popup_selection" doc:"Scan Popup: activate on selection"`
	ScanPopupAPI       bool   `toml:"scan_popup_api" doc:"Scan Popup: activate via API (socket/pipe)"`
	ScanPopupMode      string `toml:"scan_popup_mode" doc:"Scan Popup: search mode"`
	ScanPopupMinScore  int    `toml:"scan_popup_min_score" doc:"Scan Popup: minimum score (0 to 100)"`
	ScanPopupWidth     int    `toml:"scan_popup_width" doc:"Scan Popup: window width"`
	ScanPopupHeight    int    `toml:"scan_popup_height" doc:"Scan Popup: window height"`
	ScanPopupMaxCount  int32  `toml:"scan_popup_max_count" doc:"Scan Popup: max number of windows"`
	ScanPopupHistory   bool   `toml:"scan_popup_history" doc:"Scan Popup: add to history"`

	ScanPopupHeaderIcons    bool   `toml:"scan_popup_header_icons" doc:"Scan Popup: use icons for header buttons"`
	ScanPopupHeaderTemplate string `toml:"scan_popup_header_template" doc:"Scan Popup: HTML template for header (dict name and score)"`
	ScanPopupTermsStyle     string `toml:"scan_popup_terms_style" doc:"Scan Popup: CSS style string for terms"`

	ScanPopupFontSizeFactor float64 `toml:"scan_popup_font_size_factor" doc:"Scan Popup: font size factor (relative to app)"`

	ScanPopupBypassWindowManager bool `toml:"scan_popup_bypass_window_manager" doc:"Scan Popup: bypass window manager"`

	RandomFavoritePopupIntervalSeconds int `toml:"random_favorite_popup_interval_seconds" doc:"Show a random favorite term popup every N seconds (0 to disable)"`

	Audio bool `toml:"audio" doc:"Enable audio in article"`

	AudioMPV bool `toml:"audio_mpv" doc:"Use ‘mpv‘ command for playing audio"`

	AudioDownloadTimeout time.Duration `toml:"audio_download_timeout" doc:"Timeout for downloading audio files"`

	AudioAutoPlay int `toml:"audio_auto_play" doc:"Number of audio file to auto-play, set ‘0‘ to disable."`

	AudioAutoPlayWaitBetween time.Duration `toml:"audio_auto_play_wait_between" doc:"Wait time between multiple audio files on auto-play"`

	AudioAutoPlayMinScore uint8 `toml:"audio_auto_play_min_score" doc:"Minimum score for audio auto-play"`

	AudioVolume int `toml:"audio_volume" doc:"Volume for playing audio, 0 to 100 (% multiplied by dict-specific volume)"`

	EmbedExternalStylesheet bool `toml:"embed_external_stylesheet" doc:"Embed external stylesheet/css in article"`

	ResourceHttpDownloadTimeout time.Duration `toml:"resource_http_download_timeout" doc:"Timeout for downloading http/https resources in article"`

	ColorMapping map[string]string `toml:"color_mapping" doc:"Mapping for colors used in article"`

	LoadingPopupStyleStr string `toml:"loading_popup_style_str" doc:"Stylesheet (text) for 'Loading' popup"`

	ArticleZoomFactor float64 `toml:"article_zoom_factor" doc:"Zoom factor for article with mouse wheel or keyboard"`

	ArticleArrowKeys bool `toml:"article_arrow_keys" doc:"Use arrow keys to scroll through article (when focused)"`

	MiscButtonsVerticalPadding int `toml:"misc_buttons_vertical_padding" doc:"Misc buttons vertical padding"`

	ReduceMinimumWindowWidth bool `toml:"reduce_minimum_window_width" doc:"Use smaller buttons to reduce minimum width of window"`

	LocalServerPorts []string `toml:"local_server_ports" doc:"Ports for local server. Server runs on first port; Client tries all"`

	LocalClientTimeout time.Duration `toml:"local_client_timeout" doc:"Timeout for local web client"`

	WebEnable bool `toml:"web_enable" doc:"Set true/false and restart to enable/disable web service & web app"`
	WebExpose bool `toml:"web_expose" doc:"Expose web service & web app to outside (otherwise only available to 127.0.0.1)"`

	WebSearchOnType          bool `toml:"web_search_on_type" doc:"Web: Enable/disable search-on-type"`
	WebSearchOnTypeMinLength int  `toml:"web_search_on_type_min_length" doc:"Web: Minimum query length for search-on-type"`
	WebSearchOnTypeOnRegex   bool `toml:"web_search_on_type_on_regex" doc:"Web: Enable/disable search-on-type in Regex mode"`

	WebShowPoweredBy bool `toml:"web_show_powered_by" doc:"Show 'Powered By ...' footer in web."`

	SearchWorkerCount int `toml:"search_worker_count" doc:"The number of workers / goroutines used for search"`

	SearchTimeout time.Duration `toml:"search_timeout" doc:"Timeout for search on each dictionary. Only works if ‘search_worker_count > 1‘"`
}

const defaultHeaderTemplate = `<b><font color='#55f'>{{.DictName}}</font></b>
<font color='#777'> [Score: %{{.Score}}]</font>
{{if .ShowTerms }}
<span dir="ltr" style="font-size: xx-large;font-weight:bold;">
{{ index .Terms 0 }}
</span>
{{range slice .Terms 1}}
<span dir="ltr" style="font-size: large;font-weight:bold;">
	<span style="color:#ff0000;font-weight:bold;"> │ </span>
	{{ . }}
</span>
{{end}}
{{end}}`

func Default() *Config {
	return &Config{
		Logging: LoggingConfig{
			NoColor: false,
			Level:   "info",
		},
		DirectoryList: []string{
			".stardict/dic",
		},

		Style: "",

		ArticleStyle: "",

		FontFamily: "",
		FontSize:   0,

		SearchOnType:          false,
		SearchOnTypeMinLength: 3,
		SearchOnTypeOnRegex:   false,

		HeaderTemplate: defaultHeaderTemplate,
		HeaderWordWrap: true,

		HistoryDisable:  false,
		HistoryAutoSave: true,
		HistoryMaxSize:  100,

		MostFrequentDisable:  false,
		MostFrequentAutoSave: true,
		MostFrequentMaxSize:  100,

		FavoriteButtonImage: "favorite-green-64.png",

		FavoriteButtonHue: 120,

		MaxResultsTotal: 40,

		ResultTree: false,

		RandomFavoriteSearchMode: "wordMatch",

		StartHidden: false,

		DesktopWidget:          true,
		DesktopWidgetClickTime: 100,

		ScanPopupClipboard: false,
		ScanPopupSelection: false,
		ScanPopupAPI:       true,
		ScanPopupMode:      "fuzzy",
		ScanPopupMinScore:  0,
		ScanPopupWidth:     700,
		ScanPopupHeight:    400,
		ScanPopupMaxCount:  3,
		ScanPopupHistory:   true,

		ScanPopupHeaderIcons:    true,
		ScanPopupHeaderTemplate: `<span style='font-weight:200;'>{{.DictName}} [Score: %{{.Score}}]</span>`,
		ScanPopupTermsStyle:     `style="font-size:large;font-weight:bold;"`,

		ScanPopupFontSizeFactor: 0.8,

		ScanPopupBypassWindowManager: true,

		Audio: true,

		AudioMPV: false,

		AudioDownloadTimeout: 1000 * time.Millisecond,

		AudioAutoPlay:            1,
		AudioAutoPlayWaitBetween: 800 * time.Millisecond,
		AudioAutoPlayMinScore:    100,

		AudioVolume: 70,

		EmbedExternalStylesheet: false,

		ResourceHttpDownloadTimeout: 2 * time.Second,

		ColorMapping: map[string]string{},

		LoadingPopupStyleStr: "border: 1px solid red; background-color: #333; color: white",

		ArticleZoomFactor: 1.1,

		ArticleArrowKeys: false,

		MiscButtonsVerticalPadding: 5,

		ReduceMinimumWindowWidth: false,

		LocalServerPorts: []string{
			"8357",
		},

		LocalClientTimeout: 100 * time.Millisecond,

		WebEnable: false,
		WebExpose: false,

		WebSearchOnType:          false,
		WebSearchOnTypeMinLength: 3,
		WebSearchOnTypeOnRegex:   false,

		WebShowPoweredBy: true,

		SearchWorkerCount: 8,

		SearchTimeout: 5 * time.Second,
	}
}

func Path() string {
	_path := os.Getenv("CONFIG_FILE")
	if _path != "" {
		absPath, err := filepath.Abs(_path)
		if err == nil {
			return absPath
		} else {
			slog.Error("bad CONFIG_FILE", "CONFIG_FILE", _path, "err", err)
		}
	}
	return filepath.Join(GetConfigDir(), fileName)
}

func GetConfigDir() string {
	if os.Getenv("CONFIG_FILE") != "" {
		return filepath.Dir(Path())
	}
	return Paths.ConfigDir()
}

func loadFile() ([]byte, error) {
	mutex.Lock()
	defer mutex.Unlock()

	pathStr := Path()
	tomlBytes, err := os.ReadFile(pathStr)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	return tomlBytes, nil
}

func Load() (*Config, error) {
	tomlBytes, err := loadFile()
	if err != nil {
		return nil, err
	}
	conf := Default()
	if tomlBytes != nil {
		_, err = toml.Decode(string(tomlBytes), conf)
		if err != nil {
			return nil, err
		}
	}
	if len(conf.LocalServerPorts) == 0 {
		return nil, errors.New("config local_server_ports must contain at least one port")
	}
	return conf, nil
}

func MustLoad() *Config {
	conf, err := Load()
	if err != nil {
		panic(err)
	}
	return conf
}
